frontend/src/components/widgets/FileUploader/FileUploader.tsx

```diff
@@ -64,62 +74,118 @@ class FileUploader extends React.PureComponent<Props, State> {
-        })
+    // Too many files were uploaded. Upload the first eligible file
+    // and reject the rest
+    if (rejectedFiles.length > 1 && !multipleFiles) {
+      const firstFileIndex: number = rejectedFiles.findIndex(
+        file =>
+          file.errors.length === 1 && file.errors[0].code === "too-many-files"
+      )
 
-        return
+      const parsedRejections = [...rejectedFiles]
+      if (firstFileIndex >= 0) {
+        const firstFile: FileRejection = parsedRejections.splice(
+          firstFileIndex,
+          1
+        )[0]
+        this.uploadFile(firstFile.file, acceptedFiles.length)
       }
+      this.rejectFiles(parsedRejections)
+    } else {
+      this.rejectFiles(rejectedFiles)
     }
 
-    this.setState({ acceptedFiles, status: "UPLOADING" })
+    acceptedFiles.map(this.uploadFile)
+  }
+
+  private handleFile = (file: ExtendedFile, index: number): void => {
+    // Add an unique ID to each file for server and client to sync on
+    file.id = `${index}${new Date().getTime()}`
+    // Add a cancel token to cancel file upload
+    file.cancelToken = axios.CancelToken.source()
+    this.setState(state => ({ files: [file, ...state.files] }))
+  }
 
-    // Upload all the files
-    this.currentUploadCanceller = axios.CancelToken.source()
+  private uploadFile = (file: ExtendedFile, index: number): void => {
+    file.progress = 1
+    file.status = FileStatuses.UPLOADING
+    this.handleFile(file, index)
     this.props.uploadClient
       .uploadFiles(
         this.props.element.get("id"),
-        acceptedFiles,
-        undefined,
-        this.currentUploadCanceller.token
+        [file],
+        e => this.onUploadProgress(e, file),
+        file.cancelToken
+          ? file.cancelToken.token
+          : axios.CancelToken.source().token
       )
       .then(() => {
-        this.currentUploadCanceller = undefined
-        this.setState({ status: "UPLOADED" })
+        this.setState(state => {
+          const files = state.files.map(existingFile => {
+            // Destructing a file object causes us to lose the
+            // File object properties i.e. size.
+            if (file.id === existingFile.id) {
+              delete file.progress
+              delete file.cancelToken
+              file.status = FileStatuses.UPLOADED
+              return file
+            }
+            return existingFile
+          })
+          return { files }
+        })
       })
       .catch(err => {
         if (axios.isCancel(err)) {
           // If this was a cancel error, we don't show the user an error -
           // the cancellation was in response to an action they took
-          this.currentUploadCanceller = undefined
-          this.setState({ status: "UPLOADED" })
+          const files = this.state.files.map(existingFile => {
+            if (file.id === existingFile.id) {
+              return file
+            }
+            return existingFile
+          })
+          this.setState({ files })
```

Comment:
As the variable `files` depends on the current state, I think `setState` should be called using its other form.